xrootd/mux/mux.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mux // import "go-hep.org/x/hep/xrootd/mux"
+
+import (
+	"math"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// ServerResponse contains slice of bytes Data representing data from
+// XRootD server response (see XRootD protocol specification) and
+// Err representing error received from server or occurred
+// during response decoding.
+type ServerResponse struct {
+	Data []byte
+	Err  error
+}
+
+type dataSendChan chan<- ServerResponse
+type DataRecvChan <-chan ServerResponse
+
+const streamIDPartSize = math.MaxUint8
+const streamIDPoolSize = streamIDPartSize * streamIDPartSize
+
+// Mux manages channels by their ids.
+// Basically, it's a map[StreamID] chan<-*ServerResponse
+// with methods to claim, free and pass data to a specific channel by id.
+type Mux struct {
+	mu          sync.Mutex
+	dataWaiters map[protocol.StreamID]dataSendChan
+	freeIDs     chan uint16
```

Comment:
If freeIDs is unbuffered, then benchmark result is ~700 ns /op vs ~350 ns /op before.
If it's buffered (with buffer at least 20 or 30, for example) then benchmark result is close to ~350 ns/op.

Should I make it buffered? If so, which value to choose?

Reply:
let's use the magic number of `32` then (with a comment saying it's completely arbitrary at the moment and that value should be refined based on real use cases.)